File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/CustomControls.java
package net.kdt.pojavlaunch.customcontrols;

import androidx.annotation.Keep;

import com.arata.yukarilauncher.ui.subassembly.customcontrols.ControlInfoData;

import net.kdt.pojavlaunch.Tools;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@Keep
public class CustomControls {
	public int version = 7;
    public float scaledAt;
	public List<ControlData> mControlDataList;
	public List<ControlDrawerData> mDrawerDataList;
	public List<ControlJoystickData> mJoystickDataList;
	public ControlInfoData mControlInfoDataList;
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
	public CustomControls() {
		this(new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), new ControlInfoData());
	}
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
	public CustomControls(List<ControlData> mControlDataList, List<ControlDrawerData> mDrawerDataList, List<ControlJoystickData> mJoystickDataList, ControlInfoData mControlInfoDataList) {
		this.mControlDataList = mControlDataList;
		this.mDrawerDataList = mDrawerDataList;
		this.mJoystickDataList = mJoystickDataList;
		this.mControlInfoDataList = mControlInfoDataList;
		this.scaledAt = 100f;
	}
/**
 * 「save」メソッド。
 * このクラスに定義された機能メソッドです。
 */
	public void save(String path) throws IOException {
		// 現在のバージョンはV3.2のため、バージョンは8とマークする必要があります！
		version = 8;
		Tools.write(path, Tools.GLOBAL_GSON.toJson(this));
	}
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/CustomControls.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols

import androidx.annotation.Keep
import com.arata.yukarilauncher.ui.subassembly.customcontrols.ControlInfoData
import com.arata.yukarilauncher.Tools
import java.io.IOException
import java.util.ArrayList
import java.util.List

/**
 * カスタムコントロールのレイアウト全体を表すデータクラス。
 * ボタン、ドロワー、ジョイスティックの各リストを持ち、JSONとして保存・読み込みが行われます。
 */
@Keep
class CustomControls {
    /** レイアウトファイルのバージョン。 */
    var version = 7
    /** スケール適用時の基準値。 */
    var scaledAt: Float = 0f
    /** コントロールボタンのデータリスト。 */
    var mControlDataList: MutableList<ControlData>? = null
    /** ドロワー（格納型コントロール）のデータリスト。 */
    var mDrawerDataList: MutableList<ControlDrawerData>? = null
    /** ジョイスティックのデータリスト。 */
    var mJoystickDataList: MutableList<ControlJoystickData>? = null
    /** コントロール情報のメタデータ。 */
    var mControlInfoDataList: ControlInfoData? = null

    constructor() : this(ArrayList(), ArrayList(), ArrayList(), ControlInfoData())

    /**
     * @param mControlDataList コントロールボタンデータのリスト
     * @param mDrawerDataList ドロワーデータのリスト
     * @param mJoystickDataList ジョイスティックデータのリスト
     * @param mControlInfoDataList コントロール情報データ
     */
    constructor(mControlDataList: MutableList<ControlData>, mDrawerDataList: MutableList<ControlDrawerData>, mJoystickDataList: MutableList<ControlJoystickData>, mControlInfoDataList: ControlInfoData) {
        this.mControlDataList = mControlDataList
        this.mDrawerDataList = mDrawerDataList
        this.mJoystickDataList = mJoystickDataList
        this.mControlInfoDataList = mControlInfoDataList
        this.scaledAt = 100f
    }

    /**
     * レイアウトデータをJSONファイルとして保存します。
     * @param path 保存先のファイルパス
     * @throws IOException ファイル書き込みエラー
     */
    @Throws(IOException::class)
    fun save(path: String) {
        version = 8
        Tools.write(path, Tools.GLOBAL_GSON.toJson(this))
    }
}